miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,385 @@
+use alloc::vec::Vec;
+use core::iter;
+
+use winter_utils::{Deserializable, DeserializationError, Serializable};
+
+use super::{EmptySubtreeRoots, MerklePath, RpoDigest, SMT_MAX_DEPTH, Word};
+
+/// A different representation of [`MerklePath`] designed for memory efficiency for Merkle paths
+/// with empty nodes.
+///
+/// Empty nodes in the path are stored only as their position, represented with a bitmask. A
+/// maximum of 64 nodes in the path can be empty. The number of empty nodes has no effect on memory
+/// usage by this struct, but will incur overhead during iteration or conversion to a
+/// [`MerklePath`], for each empty node.
+///
+/// NOTE: This type assumes that Merkle paths always span from the root of the tree to a leaf.
+/// Partial paths are not supported.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct SparseMerklePath {
+    /// A bitmask representing empty nodes. The set bit corresponds to the depth of an empty node.
+    empty_nodes: u64,
+    /// The non-empty nodes, stored in depth-order, but not contiguous across depth.
+    nodes: Vec<RpoDigest>,
+}
+
+impl SparseMerklePath {
+    /// Constructs a sparse Merkle path from any iterator which knows its exact size.
+    ///
+    /// Knowing the size is necessary to calculate the depth of the tree, which is needed to detect
+    /// which nodes are empty nodes.
+    pub fn from_sized_iter(iter: impl ExactSizeIterator<Item = RpoDigest>) -> Option<Self> {
+        let tree_depth: u8 = iter.len() as u8;
+        if tree_depth > SMT_MAX_DEPTH {
+            return None;
+        }
+
+        let mut nodes: Vec<RpoDigest> = Default::default();
+        let mut empty_nodes: u64 = 0;
+
+        for (depth, node) in iter::zip(0..tree_depth, iter) {
+            let &equivalent_empty_node = EmptySubtreeRoots::entry(tree_depth, depth);
+            if node == equivalent_empty_node {
+                // FIXME: should we just fallback to the Vec if we're out of bits?
+                assert!(depth < 64, "SparseMerklePath may have at most 64 empty nodes");
+                empty_nodes |= u64::checked_shl(1, depth.into()).unwrap();
+            } else {
+                nodes.push(node);
+            }
+        }
+
+        Some(Self { empty_nodes, nodes })
+    }
+
+    /// Converts a Merkle path to a sparse representation.
+    ///
+    /// Returns `None` if `path` contains more elements than we can represent ([`SMT_MAX_DEPTH`]).
+    pub fn from_path(path: MerklePath) -> Option<Self> {
+        // Note that the path does not include the node itself that it is a path to.
+        // That is to say, the path is not inclusive of its ending.
+
+        Self::from_sized_iter(path.into_iter())
+    }
+
+    /// Converts this sparse representation back to a normal [`MerklePath`].
+    pub fn into_path(mut self) -> MerklePath {
```

Comment:
I would probably get rid of this function and move its code into `impl From<SparseMerklePath> for MerklePath`.